### PR TITLE
Support PointRend

### PR DESCRIPTION
**Pull request by xvjiarui**

### CODE REVIEW

File: mmdet/models/roi_heads/point_heads/point_head.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        self.in_channels = in_channels
+        self.fc_channles = fc_channels
+        self.num_classes = num_classes
+        self.class_agnostic = class_agnostic
+        self.coarse_pred_each_layer = coarse_pred_each_layer
+        self.conv_cfg = conv_cfg
+        self.norm_cfg = norm_cfg
+        self.loss_point = build_loss(loss_point)
+
+        fc_channels_in = in_channels + num_classes
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_heads/point_head.py:39]: fc_channels_in -> fc_in_channels

File: mmdet/ops/point_sample.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+def denormalize(grid):
+    """
+    convert from range [0, 1] to [-1, 1]
+    """
+    return grid * 2.0 - 1.0
+
+
+def generate_grid(num_grid, size, device):
+    """
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:24]: Docstring

File: mmdet/ops/point_sample.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    affine_trans = torch.tensor([[[1., 0., 0.], [0., 1., 0.]]], device=device)
+    grid = F.affine_grid(
+        affine_trans, torch.Size((1, 1, *size)), align_corners=False)
+    grid = normalize(grid)
+    return grid.view(1, -1, 2).expand(num_grid, -1, -1)
+
+
+def rel_roi_point2abs_img_point(rois, rel_roi_points):
+    """ Convert roi based relative point coordinates to image based absolute
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:44]:
```python
"""Convert...
```

File: mmdet/ops/point_sample.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            rois[:, None, 2] - rois[:, None, 0])
+        abs_img_points[:, :, 1] = abs_img_points[:, :, 1] * (
+            rois[:, None, 3] - rois[:, None, 1])
+        abs_img_points[:, :, 0] += rois[:, None, 0]
+        abs_img_points[:, :, 1] += rois[:, None, 1]
+    return abs_img_points
+
+
+def abs_img_point2rel_img_point(abs_img_points, img_shape, spatial_scale=1.):
+    """ Convert image based absolute point coordinates to image based relative
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:74]: Docstring

File: mmdet/ops/point_sample.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    rel_img_points = abs_img_points / scale * spatial_scale
+
+    return rel_img_points
+
+
+def rel_roi_point2rel_img_point(rois,
+                                rel_roi_points,
+                                img_shape,
+                                spatial_scale=1.):
+    """ Convert roi based relative point coordinates to image based absolute
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:102]: Docstring

File: mmdet/ops/point_sample.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        input, denormalize(points), align_corners=align_corners, **kwargs)
+    if add_dim:
+        output = output.squeeze(3)
+    return output
+
+
+class SimpleRoIAlign(nn.Module):
+
+    def __init__(self, out_size, spatial_scale, aligned=True):
+        """
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:153]: May describe what it is used for and why we need this module.

File: mmdet/models/roi_heads/mask_heads/coarse_mask_head.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+from mmcv.cnn import ConvModule, constant_init, xavier_init
+
+from mmdet.core import auto_fp16
+from mmdet.models.builder import HEADS
+from .fcn_mask_head import FCNMaskHead
+
+
+@HEADS.register_module
+class CoarseMaskHead(FCNMaskHead):
+
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/mask_heads/coarse_mask_head.py:11]: Docstring, describe the main difference between `CoarseMaskHead` and `FCNMaskHead` and its usage.

File: mmdet/models/roi_heads/point_heads/point_head.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+from mmdet.models.builder import HEADS, build_loss
+from mmdet.ops import point_sample, rel_roi_point2rel_img_point
+
+
+@HEADS.register_module
+class PointHead(nn.Module):
+    """A mask point head with shared multi-layer perceptron (equivalent to
+    nn.Conv1d)
+    """
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_heads/point_head.py:15]: Use standard docstrings

**Comment by zehuichen123** [comment]
Hi, I am trying this version of pointrend on my own dataset but got `RuntimeError: grid_sampler(): expected input and grid to have same dtype, but input has c10::Half and grid has float` with fp16 training.

**Comment by xvjiarui** [comment]
> Hi, I am trying this version of pointrend on my own dataset but got `RuntimeError: grid_sampler(): expected input and grid to have same dtype, but input has c10::Half and grid has float` with fp16 training.

Hi @zehuichen123 
FP16 is not supported in PointRend yet.

File: mmdet/models/roi_heads/point_heads/point_head.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        Sample points in [0, 1] x [0, 1] coordinate space based on their
+        uncertainty. The uncertainties are calculated for each point using
+        'uncertainty_func' function that takes point's logit prediction as
+        input.
+
+        Args:
+            mask_pred (Tensor): A tensor of shape (N, C, mask_height,
+                mask_width) for class-specific or class-agnostic prediction.
+            labels (list): A list of length R that contains either predicted
+                of ground truth class for each predicted mask.
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_heads/point_head.py:178]: Add en empty line.

File: mmdet/models/roi_heads/point_heads/point_head.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        uncertainty_map = self._get_uncertainty(mask_pred, pred_label)
+        num_rois, _, mask_height, mask_width = uncertainty_map.shape
+        h_step = 1.0 / mask_height
+        w_step = 1.0 / mask_width
+
+        uncertainty_map = uncertainty_map.view(num_rois,
+                                               mask_height * mask_width)
+        num_points = min(mask_height * mask_width, num_points)
+        point_indices = uncertainty_map.topk(num_points, dim=1)[1]
+        point_coords = torch.zeros(
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_heads/point_head.py:244]: we can use `uncertainty_map.new_zeros()`?

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        # image shape of the first image in the batch (only one)
+        ori_shape = img_metas[0]['ori_shape']
+        scale_factor = img_metas[0]['scale_factor']
+        if det_bboxes.shape[0] == 0:
+            segm_result = [[] for _ in range(self.mask_head.num_classes)]
+        else:
+            # if det_bboxes is rescaled to the original image size, we need to
+            # rescale it back to the testing scale to obtain RoIs.
+            if rescale and not isinstance(scale_factor, float):
+                scale_factor = torch.from_numpy(scale_factor).to(
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/point_rend_roi_head.py:140]: `scale_factor = det_bboxes.new_tensor(scale_factor)`

File: mmdet/models/detectors/point_rend.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+from ..builder import DETECTORS
+from .two_stage import TwoStageDetector
+
+
+@DETECTORS.register_module()
+class PointRend(TwoStageDetector):
+
```
Comment (ZwwWayne) [inline mmdet/models/detectors/point_rend.py:7]: We may add the link of paper here.

File: mmdet/ops/point_sample.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
+# Modified from https://github.com/facebookresearch/detectron2/tree/master/projects/PointRend  # noqa
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.modules.utils import _pair
+
+
+def normalize(grid):
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:9]: normalize -> denormalize_grid

File: mmdet/ops/point_sample.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+    """Normalize input grid from [-1, 1] to [0, 1]
+
+    Args:
+        grid (Tensor): The grid to be normalize, range [-1, 1].
+
+    Returns:
+        Tensor: Denormalized grid, range [0, 1].
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:16]: Either line 16 and line 28 has is inaccurate and needs to be modified.

File: mmdet/ops/point_sample.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    """Generate regular square grid of points in [0, 1] x [0, 1] coordinate
+    space.
+
+    Args:
+        num_grid (int): The number of grids to sample, one for each region.
+        size (tuple(int, int)): The side size of the regular grid.
+        device (torch.device): Desired device of returned tensor.
+
+    Returns:
+        (Tensor): A tensor of shape (num_grid, size[0]*size[1], 2) that
```
Comment (ZwwWayne) [inline mmdet/ops/point_sample.py:44]: `torch.Tensor: A tensor xxxxx`

File: mmdet/models/roi_heads/point_heads/point_head.py [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        """
+        if mask_pred.shape[1] == 1:
+            gt_class_logits = mask_pred.clone()
+        else:
+            inds = torch.arange(mask_pred.shape[0], device=mask_pred.device)
+            gt_class_logits = mask_pred[inds, labels].unsqueeze(1)
+        return -torch.abs(gt_class_logits)
+
+    def get_roi_rel_points_train(self, mask_pred, labels, cfg):
+        """
```
Comment (hellock) [inline mmdet/models/roi_heads/point_heads/point_head.py:168]: A summary is missing.

File: mmdet/models/roi_heads/point_rend_roi_head.py [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    def __init__(self, point_head, *args, **kwargs):
+        super(PointRendRoIHead, self).__init__(*args, **kwargs)
+        assert self.with_bbox and self.with_mask
+        self.init_point_head(point_head)
+
+    def init_point_head(self, point_head):
+        self.point_head = builder.build_head(point_head)
+
+    def init_weights(self, pretrained):
+        super(PointRendRoIHead, self).init_weights(pretrained)
```
Comment (hellock) [inline mmdet/models/roi_heads/point_rend_roi_head.py:27]: `super().init_weights(pretrained)`

File: configs/point_rend/README.md [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+```
+@InProceedings{kirillov2019pointrend,
+  title={{PointRend}: Image Segmentation as Rendering},
+  author={Alexander Kirillov and Yuxin Wu and Kaiming He and Ross Girshick},
+  journal={ArXiv:1912.08193},
+  year={2019}
+}
+```
+
+## Results and models
```
Comment (hellock) [inline configs/point_rend/README.md:13]: Indicate multi-scale training.

**Pull request merged**